File: python/word-search/word_search.py
class Point:
    def __init__(self, x, y):
        self.x = x  # Initialize the x-coordinate of the Point object
        self.y = y  # Initialize the y-coordinate of the Point object

    def __eq__(self, other):
        # Define the equality comparison method for Point objects
        return self.x == other.x and self.y == other.y


class WordSearch:
    def __init__(self, puzzle):
        self.puzzle = puzzle

    def search(self, word):
        word = word.lower()  # Convert the word to lowercase for case-insensitive search

        # Define possible directions (horizontal, vertical, diagonal, and reverse diagonal)
        directions = [
            (0, 1),  # Right
            (1, 0),  # Down
            (1, 1),  # Diagonal down-right
            (-1, 1),  # Diagonal down-left
        ]

        for row in range(len(self.puzzle)):
            for col in range(len(self.puzzle[row])):
                for dx, dy in directions:
                    x, y = col, row  # Initialize starting point
                    found = True  # Assume the word is found
                    for letter in word:
                        if (
                            x < 0
                            or x >= len(self.puzzle[0])
                            or y < 0
                            or y >= len(self.puzzle)
                            or self.puzzle[y][x].lower() != letter
                        ):
                            found = False  # Word not found
                            break
                        x += dx  # Move in the current direction
                        y += dy
                    if found:
                        # Calculate the starting and ending points of the word
                        start = Point(col, row)
                        end = Point(x - dx, y - dy)
                        return start, end  # Return the first occurrence

        # Return None if the word is not found
        return None

File: javascript/word-search/word-search.js
class WordSearch {
  // This class represents a word search puzzle.
  constructor(grid) {
    // The grid is a 2D array of characters that represents the puzzle.
    this.grid = grid;
    // The possible directions in which to search for words.
    this.possibleDirections = [
      { horizontal: 0, vertical: 1 },
      { horizontal: 0, vertical: -1 },
      { horizontal: 1, vertical: 0 },
      { horizontal: -1, vertical: 0 },
      { horizontal: 1, vertical: 1 },
      { horizontal: -1, vertical: 1 },
      { horizontal: 1, vertical: -1 },
      { horizontal: -1, vertical: -1 }
    ];
  }

  // This method finds all of the words in the given list that are present in the puzzle.
  find(wordsToSearch) {
    // The wordsLeft array contains the words that have not yet been found.
    const wordsLeft = [...wordsToSearch];
    // The foundWords object stores the coordinates of the found words.
    const foundWords = {};
    // Iterate over each row and column in the grid.
    for(let row = 0; row < this.grid.length; row++) {
      for(let column = 0; column < this.grid[row].length; column++) {
        // Iterate over each word in the wordsLeft array.
        wordsLeft.forEach(word => {
          // Iterate over each possible direction.
          for(const direction of this.possibleDirections) {
            // Get the coordinates of the word in the given direction.
            const wordCoordenates = this.getWordCoordenates(word, direction, { row, column });
            // If the word is found, add it to the foundWords object.
            if(wordCoordenates) {
              foundWords[word] = wordCoordenates; 
            }
          }
        });
      }
    }
    // Return the foundWords object.
    return foundWords;
  }
  
  // This method gets the coordinates of a word in the given direction.
  getWordCoordenates(word, direction, initialCoordenate) {
    // Iterate over each character in the word.
    for(let i = 0; i < word.length; i++) {
      // Get the row and column of the current character.
      const row = initialCoordenate.row + direction.vertical * i;
      const column = initialCoordenate.column + direction.horizontal * i;
      // If the character does not match the corresponding character in the word, return undefined.
      if(this.grid[row]?.[column] !== word[i]) {
        return;
      }
    }
    // Return the coordinates of the word.
    return { 
      start: [initialCoordenate.row + 1, initialCoordenate.column + 1],
      end: [
        initialCoordenate.row + 1 + (word.length - 1) * direction.vertical,
        initialCoordenate.column + 1 + (word.length - 1) * direction.horizontal
      ]
    };
  }
}

export default WordSearch;
